Readme.md

# autod

[![NPM version][npm-image]][npm-url]
[![David deps][david-image]][david-url]
[![node version][node-image]][node-url]
[![Gittip][gittip-image]][gittip-url]

[npm-image]: https://img.shields.io/npm/v/autod.svg?style=flat-square
[npm-url]: https://npmjs.org/package/autod
[david-image]: https://img.shields.io/david/node-modules/autod.svg?style=flat-square
[david-url]: https://david-dm.org/node-modules/autod
[node-image]: https://img.shields.io/badge/node.js-%3E=_0.10-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/
[gittip-image]: https://img.shields.io/gittip/dead-horse.svg?style=flat-square
[gittip-url]: https://www.gittip.com/dead-horse/

Auto generate dependencies and devDependencies by parse the project file.
`autod` will parse all the js files in `path`, and get the latest dependencies version from [registry.npmjs.org](https://registry.npmjs.org) or other registries by `-r`.

## install

```bash
$ npm install -g autod
```

## usage

```bash
$ bin/autod -h

 Usage: autod [options]

  Options:

    -h, --help                                           output usage information
    -V, --version                                        output the version number
    -p, --path [root path]                               the root path to be parse
    -t, --test <test/benchmark/example directory paths>  modules in these paths will be tread as devDependencies
    -e, --exclude <exclude directory path>               exclude parse directory, split by `,`
    -r, --registry <remote registry>                     get latest version from which registry
    -f, --prefix [version prefix]                        version prefix, can be `~` or `^`
    -F, --devprefix [dev dependencies version prefix]    develop dependencies version prefix, can be `~` or `^`
    -w, --write                                          write dependencies into package.json
    -i, --ignore                                         ignore errors, display the dependencies or write the dependencies.
    -m, --map                                            display all the dependencies require by which file
    -d, --dep <dependently modules>                      modules that not require in source file, but you need them as dependencies
    -D, --devdep <dev dependently modules>               modules that not require in source file, but you need them in as devDependencies
    -k, --keep <dependently modules>                     modules that you want to keep version in package.json file
    -s, --semver <dependencies@version>                  auto update these modules within the specified semver
    --depignore <modules>                                modules that you want to ignore, it useful in modules alias or lerna project
```

* Autod will parse all the js files in `path`, and you can exclude folder by `-e, --exclude`.
* All the modules in test folder (can be alter by `-t, --text`) will parsed as devDependencies.
* If you set `-w, --write`, `autod` will write the dependencies into package.json file. `dependencies` will replace `dependencies` in package.json, and `devDependencies` will merge with `devDependencies` in package.json, then write into package file.
* `-f, --prefix` will add prefix to each dependencies' version.
* `-F, --devprefix` will add prefix to each dev dependencies' version.
* `-i, --ignore` will display or wrtie the dependencies even some error happened.
* `-d --dep` will add modules to dependencies even not require by any file.
* `-D --devdep` will add modules to devDependencies even not require by any file.
* `-k --keep` will keep the modules' version in package.json not change by autod.
* `-s, --semver` will update these modules within the specified semver.
* `-n, --notransform` disable transfrom es next, don't support es6 modules.

a simple example of autod:

```
autod -w --prefix="~" -d connect -D mocha,should -k express -s connect@2
```

## Maintains your dependencies in Makefile

add a command in your Makefile

```sh
autod:
    @./node_modules/.bin/autod -w
    @npm install

```

then run `make autod`, it will find all the dependencies and devDependencies in your project,
add / remove dependencies, and bump the versions.

check out some examples:

 - [cnpmjs.org](https://github.com/cnpm/cnpmjs.org/blob/master/Makefile#L95)
 - [koa-generic-session](https://github.com/koajs/generic-session/blob/master/Makefile#L40)

## es6 modules support

All files will compiled by `babel` with `babel-preset-react`, `babel-preset-es2015`, `babel-preset-stage-0`.

## Plugin support

You can write a plugin for autod to decide how to parse dependencies.

- Write a plugin, and publish it to NPM.

```js
module.exports = function (filepath, content) {
  // find the dependencies
  return []; // return dependencies
};
```

- Use plugin:

install plugin from NPM, then use it with `autod`:

```
autod -P pluginName
```

## Config support

You can put all the options in `${cwd}/.autod.conf.js`, when you run `autod`, it will load this file as input options. It can be both a `js` file or a `json` file.

```js
module.exports = {
  write: true,
  prefix: '~',
  devprefix: '^',
  dep: [
    'bluebird'
  ],
  semver: [
    'koa-router@4',
    'koa-compose@2'
  ],
};
```

## License

(The MIT License)

Copyright (c) 2013~2015 dead_horse <dead_horse@qq.com>;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

3.1.2 / 2021-11-15
==================

**others**
  * [[`cdf99e8`](http://github.com/node-modules/autod/commit/cdf99e80f8ca61b0d3f71bb6fde116f445fc4c92)] - chore: update taobao npm registry url (#50) (Janlay <<36876080+Janlaywss@users.noreply.github.com>>)

3.1.1 / 2020-11-03
==================

**fixes**
  * [[`c33cf61`](http://github.com/node-modules/autod/commit/c33cf616007d546ce0848a6efd60769abe310626)] - fix: path compatibility at windows (#48) (weijiafu14 <<1527752351@qq.com>>)

**others**
  * [[`1a3994a`](http://github.com/node-modules/autod/commit/1a3994a4b8ea63dd802f26235c15e0f2be7c6aa6)] - chore: GitHub Actions (#47) (weijiafu14 <<1527752351@qq.com>>)

3.1.0 / 2019-03-28
==================

**features**
  * [[`82db2fd`](http://github.com/node-modules/autod/commit/82db2fd1e84130e76a436102733982e536876822)] - feat: pass modules to plugin (gliudong <<gliudong@users.noreply.github.com>>)

**fixes**
  * [[`0864e9a`](http://github.com/node-modules/autod/commit/0864e9a942a4180b888c7ece3f9492084f086bc6)] - fix: range must parse in local (#39) (Yiyu He <<dead_horse@qq.com>>)
  * [[`87614ad`](http://github.com/node-modules/autod/commit/87614ad507afcaaca6bc8ce989b6ac8ee4a2d7cb)] - fix: should ignore vscode config (#36) (TZ | 天猪 <<atian25@qq.com>>)

3.0.1 / 2017-11-13
==================

**fixes**
  * [[`fb286d6`](http://github.com/node-modules/autod/commit/fb286d6f59ec5d6107cbe28d56bca37a0e01ef99)] - fix: devDependencies in dependencies should pass autod check (#31) (Yiyu He <<dead_horse@qq.com>>)

**others**
  * [[`b470681`](http://github.com/node-modules/autod/commit/b470681215df58441e83bfe00c428e9bcc92bb53)] - chore: clean (dead-horse <<dead_horse@qq.com>>)

3.0.0 / 2017-11-11
==================

**fixes**
  * [[`eca5d8f`](http://github.com/node-modules/autod/commit/eca5d8fc70d8d704b36446c0cf29ecc28b2239fe)] - fix: don't compare if no import (dead-horse <<dead_horse@qq.com>>)

**others**
  * [[`b752ed2`](http://github.com/node-modules/autod/commit/b752ed27552356b719d638d32e0ef24739d990f9)] - deps: update dependencies (dead-horse <<dead_horse@qq.com>>)
  * [[`960f13a`](http://github.com/node-modules/autod/commit/960f13a5625cadc92ba5f557f51c50846418f360)] - refactor: rewrite core logic, speed up! (Yiyu He <<dead_horse@qq.com>>)

2.10.1 / 2017-10-12
==================

**fixes**
  * [[`16e401b`](http://github.com/node-modules/autod/commit/16e401b2e2ccafc9471cab3cfcc525d98d42fdc8)] - fix: dont write when check (dead-horse <<dead_horse@qq.com>>)

2.10.0 / 2017-10-12
==================

**features**
  * [[`e142b00`](http://github.com/node-modules/autod/commit/e142b00931a36b3c3a5c8a30b37dfa2f31a01725)] - feat: check don't output dependencies info (dead-horse <<dead_horse@qq.com>>)

**fixes**
  * [[`8f769c3`](http://github.com/node-modules/autod/commit/8f769c37526f5ca72c5f52262ffab9367519dac1)] - fix: use preset-env (dead-horse <<dead_horse@qq.com>>)

2.9.0 / 2017-07-31
==================

**features**
  * [[`c7c309d`](http://github.com/node-modules/autod/commit/c7c309da0729cb81ef77af4c429d7f7877f24952)] - feat: support autod --check (#28) (fengmk2 <<fengmk2@gmail.com>>)

2.8.0 / 2017-03-22
==================

  * feat: support require.resolve (#27)

2.7.1 / 2016-08-19
==================

  * fix: keep will keep module not exist

2.7.0 / 2016-08-03
==================

  * deps: use ^
  * chore(deps): upgrade babel-core (#22)

2.6.1 / 2016-06-04
==================

  * fix: fix test options

2.6.0 / 2016-04-28
==================

  * feat: add cov and coverage to default exclude dir
  * fix: exclude test roots

2.5.2 / 2016-04-19 
==================

  * fix: support all es2015 and stage0
  * fix: remove registry url slash (#21)

2.5.1 / 2016-04-09
==================

  * fix: require preset and plugin

2.5.0 / 2016-04-09
==================

  * feat: support jsx
  * fix: use https://registry.npmjs.org
  * feat: read env.npm_config_registry

2.4.2 / 2016-01-21
==================

  * fix: 修复 semver

2.4.1 / 2015-12-23
==================

  * fix: fix keep

2.4.0 / 2015-12-23
==================

  * feat: support .autod.conf.js

2.3.0 / 2015-12-12
==================

  * feat: support autod plugin

2.2.0 / 2015-12-08
==================

  * desp: update dependencies
  * feat: support diable transfrom

2.1.3 / 2015-09-17
==================

  * traceur instead of babel

2.1.2 / 2015-09-17
==================

  * hotfix require

2.1.1 / 2015-09-14
==================

  * rename

2.1.0 / 2015-09-14
==================

  * support es6 import
  * deps: upgrade semver@2.3.0

2.0.1 / 2015-02-16
==================

  * fix: fix exclude in test folder

2.0.0 / 2015-01-15
==================

  * feat: use crequire to parse the require tag, fixes #3

1.2.0 / 2015-01-13
==================

  * feat: support --devprefix
  * bump dependencies

1.1.4 / 2014-11-29
==================

  * fix indent.

1.1.3 / 2014-11-29
==================

  * better display for print dependency map

1.1.2 / 2014-11-20
==================

  * fix typo error

1.1.1 / 2014-10-20
==================

  * Merge pull request #14 from node-modules/fix-publish-config-registry
  * fix publishConfig.registry not work bug

1.1.0 / 2014-10-18
==================

  * support array test roots

1.0.0 / 2014-10-01
==================

  * support -s for semver, close #13
  * add -D options, close #12

0.3.2 / 2014-08-06
==================

  * fix get registry from package.json

0.3.1 / 2014-08-06
==================

  * fix publishConfig.registry

0.3.0 / 2014-08-06
==================

  * Merge pull request #11 from node-modules/default-registry
  * try to read default registry from publishConfig

0.2.5 / 2014-08-04
==================

  * fix stupid typo

0.2.4 / 2014-07-27
==================

  * remove process.exit(0), fix --map

0.2.3 / 2014-07-22
==================

  * support scope

0.2.2 / 2014-06-26
==================

  * update reg to support kissy/lib/xtemplate type
  * update examples

test/autod.test.js
'use strict';

const coffee = require('coffee');
const assert = require('assert');
const path = require('path');

const autod = require.resolve('../bin/autod.js');

describe('autod', () => {
  it('should output when dependencies missing', () => {
    const cwd = path.join(__dirname, 'fixtures/check-pkg');
    return coffee.fork(autod, [ '--prefix=^' ], { cwd })
      .debug()
      .expect('code', 0)
      .expect('stdout', /debug\s+-\s+\^.*/)
      .expect('stdout', /urllib\s+-/)
      .end();
  });

  it('should output when dependencies need upgrade', () => {
    const cwd = path.join(__dirname, 'fixtures/upgrade');
    return coffee.fork(autod, [ '--prefix=^' ], { cwd })
      .debug()
      .expect('code', 0)
      .expect('stdout', /debug\s+1\s+\^.*/)
      .end();
  });

  it('should support import', () => {
    const cwd = path.join(__dirname, 'fixtures/import');
    return coffee.fork(autod, [ '--prefix=^' ], { cwd })
      .debug()
      .expect('code', 0)
      .expect('stdout', /debug\s+-\s+\^.*/)
      .end();
  });

  it('should support add test dir', () => {
    const cwd = path.join(__dirname, 'fixtures/check-pkg');
    return coffee.fork(autod, [ '--prefix=^', '--test=index.js' ], { cwd })
      .debug()
      .expect('code', 0)
      .expect('stdout', /nothing to update in Dependencies/)
      .expect('stdout', /debug\s+-\s+\^.*/)
      .end();
  });

  it('should support add exclude dir', function* () {
    const cwd = path.join(__dirname, 'fixtures/exclude');
    const res = yield coffee.fork(autod, [ '--prefix=^', '--exclude=**/fixtures' ], { cwd })
      .debug()
      .expect('code', 0)
      .end();
    assert(!res.stdout.match(/egg/));
  });

  it('should support add depignore modules', function* () {
    const cwd = path.join(__dirname, 'fixtures/depignore');
    const res = yield coffee.fork(autod, [ '--prefix=^', '--depignore=umi' ], { cwd })
      .debug()
      .expect('code', 0)
      .end();
    assert(res.stdout.match(/egg/));
    assert(!res.stdout.match(/umi/));
  });

  it('should support add semver', () => {
    const cwd = path.join(__dirname, 'fixtures/check-pkg');
    return coffee.fork(autod, [ '--prefix=^', '--semver=debug@1' ], { cwd })
      .debug()
      .expect('code', 0)
      .expect('stdout', /debug\s+-\s+\^1.*/)
      .expect('stdout', /urllib\s+-/)
      .end();
  });

  it('should support add keep', function* () {
    const cwd = path.join(__dirname, 'fixtures/check-pkg');
    const res = yield coffee.fork(autod, [ '--prefix=^', '--keep=debug' ], { cwd })
      .debug()
      .expect('code', 0)
      .end();
    assert(!res.stdout.match(/debug/));
  });
});


test/check.test.js
'use strict';

const coffee = require('coffee');
const path = require('path');

const autod = require.resolve('../bin/autod.js');

describe('autod --check', () => {
  it('should exit code 1 when dependencies missing', () => {
    const cwd = path.join(__dirname, 'fixtures/check-pkg');
    return coffee.fork(autod, [ '--check' ], { cwd })
      .debug()
      .expect('code', 1)
      .expect('stderr', /debug/)
      .expect('stderr', /urllib/)
      .end();
  });

  it('should exit code 0 when dependencies exists', () => {
    const cwd = path.join(__dirname, 'fixtures/check-pkg-pass');
    return coffee.fork(autod, [ '--check' ], { cwd })
      .debug()
      .expect('code', 0)
      .end();
  });

  it('should exit code 0 when devDependencies in dependencies', () => {
    const cwd = path.join(__dirname, 'fixtures/check-dev-pkg-pass');
    return coffee.fork(autod, [ '--check' ], { cwd })
      .debug()
      .expect('code', 0)
      .end();
  });
});


test/lib/autod.test.js
/*!
 * autod - lib/autod.js
 * Copyright(c) 2013 
 * Author: dead_horse <undefined>
 */

'use strict';

/**
 * Module dependencies.
 */
var mm = require('mm');
var should = require('should_123');


lib/autod.js
'use strict';

const debug = require('debug')('autod');
const assert = require('assert');
const glob = require('glob');
const path = require('path');
const fs = require('fs');
const readdir = require('fs-readdir-recursive');
const crequire = require('crequire');
const EventEmitter = require('events');
const co = require('co');
const urllib = require('urllib');

const DEFAULT_EXCLUDE = [ '.git', 'cov', 'coverage' ];
const DEFAULT_TEST = [ 'test', 'tests', 'test.js', 'benchmark', 'example', 'example.js' ];
const USER_AGENT = `autod@${require('../package').version} ${urllib.USER_AGENT}`;
const MODULE_REG = /^(@[0-9a-zA-Z\-\_][0-9a-zA-Z\.\-\_]*\/)?([0-9a-zA-Z\-\_][0-9a-zA-Z\.\-\_]*)/;


class Autod extends EventEmitter {
  constructor(options) {
    super();
    this.options = Object.assign({}, options);
    this.prepare();
  }

  prepare() {
    const options = this.options;
    assert(options.root, 'options.root required');
    // default options
    options.semver = options.semver || {};
    options.registry = options.registry || 'https://registry.npm.taobao.org';
    options.registry = options.registry.replace(/\/?$/, '');
    options.dep = options.dep || [];
    options.devdep = options.devdep || [];
    options.root = path.resolve(this.options.root);
    if (options.plugin) {
      try {
        const pluginPath = path.join(options.root, 'node_modules', options.plugin);
        options.plugin = require(pluginPath);
      } catch (err) {
        throw new Error(`plugin ${options.plugin} not exist!`);
      }
    }

    // parse exclude and test
    const exclude = (options.exclude || []).concat(DEFAULT_EXCLUDE);
    const test = (options.test || []).concat(DEFAULT_TEST);
    options.exclude = [];
    options.test = [];
    exclude.forEach(e => {
      options.exclude = options.exclude.concat(glob.sync(path.join(options.root, e)));
    });
    test.forEach(t => {
      options.test = options.test.concat(glob.sync(path.join(options.root, t)));
    });

    // store dependencies appear in which files
    this.dependencyMap = {};
    // store fetch npm error message
    this.errors = [];
    debug('autod inited with root: %s, exclude: %j, test: %j', options.root, options.exclude, options.test);
  }

  findJsFile() {
    const files = readdir(this.options.root, (name, index, dir) => {
      const fullname = path.join(dir, name);
      // ignore all node_modules
      if (fullname.indexOf('/node_modules/') >= 0) return false;
      // ignore specified exclude directories or files
      if (this._contains(fullname, this.options.exclude)) return false;

      if (fs.statSync(fullname).isDirectory()) return true;
      const extname = path.extname(name);
      if (extname !== '.js' && extname !== '.jsx') return false;
      return true;
    });

    const jsFiles = [];
    const jsTestFiles = [];
    files.forEach(file => {
      file = path.join(this.options.root, file);
      if (this._contains(file, this.options.test)) jsTestFiles.push(file);
      else jsFiles.push(file);
    });
    debug('findJsFile jsFiles(%j), jsTestFiles(%j)', jsFiles, jsTestFiles);
    return {
      jsFiles, jsTestFiles,
    };
  }

  findDependencies() {
    const files = this.findJsFile();
    const dependencies = new Set();
    const devDependencies = new Set();

    // add to dependencies set
    files.jsFiles.forEach(file => {
      const modules = this._getDependencies(file);
      modules.forEach(module => dependencies.add(module));
    });
    (this.options.dep || []).forEach(dev => {
      dependencies.add(dev);
    });

    // exclude dependencies, add to devDependencies set
    files.jsTestFiles.forEach(file => {
      const modules = this._getDependencies(file);
      modules.forEach(module => {
        if (!dependencies.has(module)) devDependencies.add(module);
      });
    });
    (this.options.devdep || []).forEach(dev => {
      if (!dependencies.has(module)) devDependencies.add(dev);
    });

    return {
      dependencies: Array.from(dependencies),
      devDependencies: Array.from(devDependencies),
    };
  }

  * findVersions() {
    const allDependencies = this.findDependencies();
    let versions = {};
    allDependencies.dependencies.forEach(name => {
      versions[name] = this._fetchVersion(name);
    });
    allDependencies.devDependencies.forEach(name => {
      versions[name] = this._fetchVersion(name);
    });
    versions = yield versions;
    const dependencies = {};
    const devDependencies = {};
    allDependencies.dependencies.forEach(name => {
      dependencies[name] = versions[name];
    });
    allDependencies.devDependencies.forEach(name => {
      devDependencies[name] = versions[name];
    });
    return { dependencies, devDependencies };
  }

  * _fetchVersion(name) {
    try {
      const tag = this.options.semver.hasOwnProperty(name)
        ? this.options.semver[name]
        : 'latest';
      const url = `${this.options.registry}/${name}/${tag}`;
      const res = yield urllib.request(url, {
        headers: { 'user-agent': USER_AGENT },
        gzip: true,
        timeout: 10000,
        dataType: 'json',
      });
      if (res.status !== 200) {
        throw new Error(`request ${url} response status ${res.status}`);
      }
      const version = res.data && res.data.version;
      if (!version) {
        throw new Error(`no match remote version for ${name}@${tag}`);
      }
      return version;
    } catch (err) {
      this.errors.push(err);
    }
  }

  _getDependencies(filePath) {
    let file;
    try {
      file = fs.readFileSync(filePath, 'utf-8');
      if (!this.options.notransform) {
        const res = require('babel=core').transform(file, {
          presets: [ require('babel-preset-react'), require('babel-preset-env'), require('babel-preset-stage-0') ],
        });
        file = res.code;
      }
    } catch (err) {
      this.emit('warn', `Read(or transfrom) file ${filePath} error: ${err.message}`);
    }
    const modules = [];

    crequire(file, true).forEach(r => {
      const parsed = MODULE_REG.exec(r.path);
      if (!parsed) return;
      const scope = parsed[1];
      let name = parsed[2];
      if (scope) name = scope + name;
      if (this._isCoreModule(name)) return;
      modules.push(name);
      this.dependencyMap[name] = this.dependencyMap[name] || [];
      this.dependencyMap[name].push(filePath);
    });

    // support plugin parse file
    if (this.options.plugin) {
      const pluginModules = this.options.plugin(filePath, file) || [];
      pluginModules.forEach(name => {
        modules.push(name);
        this.dependencyMap[name] = this.dependencyMap[name] || [];
        this.dependencyMap[name].push(filePath);
      });
    }

    debug('file %s get modules %j', filePath, modules);
    return modules;
  }

  _contains(path, matchs) {
    for (const match of matchs) {
      if (path.startsWith(match)) return true;
    }
  }

  _isCoreModule(name) {
    let filename;
    try {
      filename = require.resolve(name);
    } catch (err) {
      return false;
    }
    return filename === name;
  }
}

Autod.prototype.findVersions = co.wrap(Autod.prototype.findVersions);

module.exports = Autod;


index.js
'use strict';

const debug = require('debug')('autod');
const assert = require('assert');
const glob = require('glob');
const path = require('path');
const fs = require('fs');
const readdir = require('fs-readdir-recursive');
const crequire = require('crequire');
const EventEmitter = require('events');
const co = require('co');
const urllib = require('urllib');
const semver = require('semver');

const DEFAULT_EXCLUDE = [ '.git', 'cov', 'coverage', '.vscode' ];
const DEFAULT_TEST = [ 'test', 'tests', 'test.js', 'benchmark', 'example', 'example.js' ];
const USER_AGENT = `autod@${require('./package').version} ${urllib.USER_AGENT}`;
const MODULE_REG = /^(@[0-9a-zA-Z\-\_][0-9a-zA-Z\.\-\_]*\/)?([0-9a-zA-Z\-\_][0-9a-zA-Z\.\-\_]*)/;


class Autod extends EventEmitter {
  constructor(options) {
    super();
    this.options = Object.assign({}, options);
    this.prepare();
  }

  prepare() {
    const options = this.options;
    assert(options.root, 'options.root required');
    // default options
    options.semver = options.semver || {};
    options.registry = options.registry || 'https://registry.npmmirror.com';
    options.registry = options.registry.replace(/\/?$/, '');
    options.dep = options.dep || [];
    options.devdep = options.devdep || [];
    options.root = path.resolve(this.options.root);
    if (options.plugin) {
      try {
        const pluginPath = path.join(options.root, 'node_modules', options.plugin);
        options.plugin = require(pluginPath);
      } catch (err) {
        throw new Error(`plugin ${options.plugin} not exist!`);
      }
    }

    // parse exclude and test
    const exclude = (options.exclude || []).concat(DEFAULT_EXCLUDE);
    const test = (options.test || []).concat(DEFAULT_TEST);
    options.exclude = [];
    options.test = [];
    exclude.forEach(e => {
      options.exclude = options.exclude.concat(glob.sync(path.join(options.root, e)).map(path.normalize));
    });
    test.forEach(t => {
      options.test = options.test.concat(glob.sync(path.join(options.root, t)).map(path.normalize));
    });

    // store dependencies appear in which files
    this.dependencyMap = {};
    // store fetch npm error message
    this.errors = [];
    debug('autod inited with root: %s, exclude: %j, test: %j', options.root, options.exclude, options.test);
  }

  findJsFile() {
    const files = readdir(this.options.root, (name, index, dir) => {
      const fullname = path.join(dir, name);
      // ignore all node_modules
      if (fullname.indexOf(`${path.sep}node_modules${path.sep}`) >= 0) return false;
      // ignore specified exclude directories or files
      if (this._contains(fullname, this.options.exclude)) return false;

      if (fs.statSync(fullname).isDirectory()) return true;
      const extname = path.extname(name);
      if (extname !== '.js' && extname !== '.jsx') return false;
      return true;
    });

    const jsFiles = [];
    const jsTestFiles = [];
    files.forEach(file => {
      file = path.join(this.options.root, file);
      if (this._contains(file, this.options.test)) jsTestFiles.push(file);
      else jsFiles.push(file);
    });
    debug('findJsFile jsFiles(%j), jsTestFiles(%j)', jsFiles, jsTestFiles);
    return {
      jsFiles, jsTestFiles,
    };
  }

  findDependencies() {
    const files = this.findJsFile();
    const dependencies = new Set();
    const devDependencies = new Set();

    // add to dependencies set
    files.jsFiles.forEach(file => {
      const modules = this._getDependencies(file);
      modules.forEach(module => dependencies.add(module));
    });
    (this.options.dep || []).forEach(dev => {
      dependencies.add(dev);
    });

    // exclude dependencies, add to devDependencies set
    files.jsTestFiles.forEach(file => {
      const modules = this._getDependencies(file);
      modules.forEach(module => {
        if (!dependencies.has(module)) devDependencies.add(module);
      });
    });
    (this.options.devdep || []).forEach(dev => {
      if (!dependencies.has(module)) devDependencies.add(dev);
    });

    return {
      dependencies: Array.from(dependencies),
      devDependencies: Array.from(devDependencies),
    };
  }

  * findVersions() {
    const allDependencies = this.findDependencies();
    let versions = {};
    allDependencies.dependencies.forEach(name => {
      versions[name] = this._fetchVersion(name);
    });
    allDependencies.devDependencies.forEach(name => {
      versions[name] = this._fetchVersion(name);
    });
    versions = yield versions;
    const dependencies = {};
    const devDependencies = {};
    allDependencies.dependencies.forEach(name => {
      dependencies[name] = versions[name];
    });
    allDependencies.devDependencies.forEach(name => {
      devDependencies[name] = versions[name];
    });
    return { dependencies, devDependencies };
  }

  * _fetchVersion(name) {
    try {
      const tag = this.options.semver.hasOwnProperty(name)
        ? this.options.semver[name]
        : 'latest';


      let url = `${this.options.registry}/${name}/${tag}`;
      let isAllVersions = false;
      // npm don't support range now
      if (semver.validRange(tag)) {
        url = `${this.options.registry}/${name}`;
        isAllVersions = true;
      }
      const res = yield urllib.request(url, {
        headers: {
          'user-agent': USER_AGENT,
          // npm will response less data
          accept: 'application/vnd.npm.install-v1+json; q=1.0, application/json; q=0.8, */*',
         },
        gzip: true,
        timeout: 10000,
        dataType: 'json',
      });
      if (res.status !== 200) {
        throw new Error(`request ${url} response status ${res.status}`);
      }
      let version;
      if (isAllVersions) {
      // match semver in local
        const versions = res.data && res.data.versions;
        if (versions) version = semver.maxSatisfying(Object.keys(versions), tag);
      } else {
        version = res.data && res.data.version;
      }
      if (!version) {
        throw new Error(`no match remote version for ${name}@${tag}`);
      }
      return version;
    } catch (err) {
      this.errors.push(err);
    }
  }

  _getDependencies(filePath) {
    let file;
    try {
      file = fs.readFileSync(filePath, 'utf-8');

      if (!this.options.notransform && file.includes('import')) {
        const res = require('babel-core').transform(file, {
          presets: [ require('babel-preset-react'), require('babel-preset-env'), require('babel-preset-stage-0') ],
        });
        file = res.code;
      }
    } catch (err) {
      this.emit('warn', `Read(or transfrom) file ${filePath} error: ${err.message}`);
    }
    const modules = [];

    crequire(file, true).forEach(r => {
      const parsed = MODULE_REG.exec(r.path);
      if (!parsed) return;
      const scope = parsed[1];
      let name = parsed[2];
      if (scope) name = scope + name;
      if (this._isCoreModule(name)) return;
      modules.push(name);
      this.dependencyMap[name] = this.dependencyMap[name] || [];
      this.dependencyMap[name].push(filePath);
    });

    // support plugin parse file
    if (this.options.plugin) {
      const pluginModules = this.options.plugin(filePath, file, modules) || [];
      pluginModules.forEach(name => {
        modules.push(name);
        this.dependencyMap[name] = this.dependencyMap[name] || [];
        this.dependencyMap[name].push(filePath);
      });
    }

    debug('file %s get modules %j', filePath, modules);
    return modules;
  }

  _contains(path, matchs) {
    for (const match of matchs) {
      if (path.startsWith(match)) return true;
    }
  }

  _isCoreModule(name) {
    let filename;
    try {
      filename = require.resolve(name);
    } catch (err) {
      return false;
    }
    return filename === name;
  }
}

Autod.prototype.findVersions = co.wrap(Autod.prototype.findVersions);

module.exports = Autod;
